src/extensions/Relude_Extensions_BoundedEnum.re

```diff
@@ -27,4 +27,185 @@ module BoundedEnumExtensions = (E: Relude_Interface.BOUNDED_ENUM) => {
       };
       go([], start, startInt);
     };
+
+  /**
+   [BoundedEnumExtensions.inverseMapEqBy] takes a equality function of type
+   [(a, a) => bool] along with a function of type [t => a], where each [a]
+   is a unique value, and returns a new function of type [a => option(t)].
+
+   [inverseMapEqBy] performs the same operations as [inverseMapEq], but
+   it accepts an equality function for [a], instead of a first-class
+   [Eq] module.
+
+   Note: since there is no way to determine the ordering of [a] without
+   a comparator function, any call to the returned function must iterate
+   through the entire list of possible values until it finds a match or
+   reaches the end of the list. Therefore, the worst-case time complexity
+   of the returned function is **linear**, and failed lookups will always
+   be worst-case.
+
+   See {!val:inverseMapOrd} or {!val: inverseMapOrdBy} for a similar function
+   that accepts a comparator for [a], and returns a reverse-lookup function
+   with a worst-case running time of O(log(n)).
+
+   Running time for staging: O(n)
+   Running time of the returned lookup function: O(n)
+   */
+  let inverseMapEqBy: type a. ((a, a) => bool, E.t => a, a) => option(E.t) =
+    (eqA, eToA) => {
+      let rec loop = (lst, maybeE, getNextE) => {
+        switch (maybeE) {
+        | None =>
+          // generate the closure to pass back to the caller:
+          let arr = Belt.List.toArray(lst);
+          let len = Belt.Array.length(arr);
+          let return = a => {
+            let rec find = (arr, i, limit) =>
+              if (i < limit) {
+                let (e__, a__) = Belt.Array.getUnsafe(arr, i);
+                if (eqA(a__, a)) {
+                  Some(e__);
+                } else {
+                  find(arr, succ(i), limit);
+                };
+              } else {
+                None;
+              };
+            find(arr, 0, len);
+          };
+          return;
+        | Some(e) =>
+          // Continue building the associative list:
+          loop([(e, eToA(e)), ...lst], getNextE(e), getNextE)
+        };
+      };
```

Comment:
Not 100% sure, but this function or something like it might already exist in https://github.com/reazen/relude/blob/master/src/extensions/Relude_Extensions_Enum.re - like a function to enumerate all the values of an `Enum`.

Also a tiny nitpick, but it might be best to not use `return` as the name of a variable - that threw me off at first glance.

Reply:
That function from `Enum` that enumerates all the values from an enum might be interchangeable here, I agree. I'll take a look and see if I can just reuse that. But I am not totally sure, because this needs to construct not just a list of `E.t`, but an associative list of `('a, E.t)`.

One reason for the current design was to keep it efficient using a tail-recursive loop, and the most efficient version of that is to return directly from one of the branches of the `switch` expression, so that there are no operations left after the recursive loop. This requires constructing the closure inside the `switch` expression directly. Although I'm not sure it's that important, and failing to reuse an existing function could result in larger bundle/binary sizes.

I think it's probably less efficient to reuse the existing function and then `zip` it into a new associative list, only to convert it into an array afterwards. Seems like a lot of allocations, IMO. But I don't have strong opinions about it.

Reply:
My point is that I'd rather put a particular algorithm in a single place and reuse it, rather than having to nest these helper functions inside other functions. If we did that all over the place, we'd have lots of implementations of every particular function to keep track of, rather than a single implementation, which is reused everywhere. If your `loop` function does something useful in the context of `Enum` or `BoundedEnum`, I would say it would be useful to add it as a new helper function to expose at the top-level and reuse it in these other functions. If your `loop` function already exists in concept somewhere else, but is a non-ideal implementation, I'd rather just reuse that function anyway, and put the onus on that reusable function to be improved. I'm not saying we need to blindly reuse functions and force ourselves to do additional allocations or extra big-O complexity - but just that we should create solid top-level functions and reuse them where appropriate, rather than re-inventing or re-working algorithms in the contexts of other functions.

Reply:
After a second look, I think it might work out nicely to reuse the function from Enum Extensions.

Reply:
Cool - sounds good. If you think there isn't a good/workable thing to reuse that is totally fair too - I think in many cases it makes sense to create super-specialized versions of functions for the sake of performance, but I just don't want to default to doing that first if we can.